core/include/detray/tools/rk_stepper.hpp

```diff
@@ -57,6 +59,9 @@ class rk_stepper final : public base_stepper<track_t, tuple_t> {
         };
 
         stepping_data _step_data;
+
+        DETRAY_HOST_DEVICE
+        void release_step_size() { _step_size = 1. * unit_constants::mm; }
```

Comment:
To the best of my understanding, I believe this should really be the distance to the next surface from the navigator and that this should be set at the beginning of every new step. We don't have other stepping constraints to be released, yet, right?

Reply:
> the distance to the next surface from the navigator

Isn't it too big?

Reply:
The purpose of this function is to make the step size the reasonable size. The step size can be very small when it tries to reach the surface so we need to scale it to normal size before starting new step

Reply:
> > the distance to the next surface from the navigator
> 
> Isn't it too big?

I thought, that's what the ```path_limit``` is for?

Reply:
> The purpose of this function is to make the step size the reasonable size. The step size can be very small when it tries to reach the surface so we need to scale it to normal size before starting new step

It can also become very small long before we reach a surface, due to the magnetic field. And then, we have to increase the step size again somehow. Let's discuss on monday, maybe :)

Reply:
If the step size is reduced significantly by the rocky magnetic field, then we should not increase them intentionally. That's what adaptive step size is meant to be. They will recover the size automatically when it gets out of the rocky magnetic field region. 

Things get different when the track tries to reach the surface because we force the step size to be small regardless of the magnatic field

Reply:
As you said `path_limit` is set to the distance to the next surface. But it should not affect `step_size`.

Reply:
Ok, this is what confused me last time and maybe I am not seeing this right: Due to line 118 ```while (!try_rk4(stepping._step_size)) {``` the adjustment of the step size only occurs, when ```error_estimate > stepping._tolerance```. Therefore, the step size scaling can never be greater then one (maybe I miscalculated?) and the step size will never be automatically increased? From the RK I know, the step size adjustment is always called and the actual integration is only called when the error is below tolerance. But maybe I am confused about the entire thing, because here it is the other way around...
Also, if we set the initial step size estimate higher than the magnetic field allows, the algorithm will reduce the step size before applying it in the integration.

Reply:
> the adjustment of the step size only occurs, when error_estimate > stepping._tolerance

Hmm you are right. It is apparent that there is an error in the rk step. Let me fix it

Reply:
From my understanding this is on purpose, because ACTS does not actually use a 'normal' RK, but a sort of guided version using the fact that thanks to the navigator the stepper always knows how far it needs to step to find a new measurement and then it just has to decide if it can do that step or if it has to reduce the step size due to integration error. In the first case we reach the surface with minimal number of steps, in the second case we navigate again. But that is only my understanding of it

Reply:
Um. I think I get it

Reply:
I am checking `EigenStepper.ipp` of ACTS  again, but I am not entirely sure if it also never increases the step size. :/

```c++
  while (true) {
    auto res = tryRungeKuttaStep(state.stepping.stepSize);
    if (!res.ok()) {
      return res.error();
    }
    if (!!res.value()) {
      break;
    }

    stepSizeScaling =
        std::min(std::max(0.25, std::pow((state.options.tolerance /
                                          std::abs(2. * error_estimate)),
                                         0.25)),
                 4.);
```

Reply:
Ok, that I don't know. It is true that especially ```stepSizeScaling =
        std::min(std::max(0.25, std::pow((state.options.tolerance /
                                          std::abs(2. * error_estimate)),
                                         0.25)),
                 4.);```
makes only sense, if you expect to also increase the step size again.

Reply:
Hmm. Maybe i have to consider removing try_rk_step and let step adjustment happen whatever the error estimate is.

Reply:
Then i fall into dilema that the reduced step size could even be super big when the initial step size is set to the distance to the next layer. Wth

Reply:
But the initial design of the RK stepper already had a ```max_step_size``` that was checked. Only that you put the distance to the next surface/volume boundary there instead of a fixed ```max_step_size```:
```stepping._step_size = std::min(stepping._step_size, es);```